htlcswitch/switch.go

```diff
@@ -0,0 +1,770 @@
+// configuration MUST be non-nil for the service to carry out its duties.
+type Config struct {
+	// LocalChannelClose kicks-off the workflow to execute a cooperative
+	// or forced unilateral closure of the channel initiated by a local
+	// subsystem.
+	LocalChannelClose func(pubKey []byte, request *ChanClose)
+}
+
+// Switch is a central messaging bus for all incoming/outgoing htlc's.
+// The goal of the switch is forward the incoming/outgoing htlc messages from
+// one channel to another, and also propagate the settle/fail htlc messages
+// back to original requester by using payment circuits. Also switch is
+// responsible for notifying the user about result of payment request.
+type Switch struct {
+	started  int32
+	shutdown int32
+	wg       sync.WaitGroup
+	quit     chan struct{}
+	mutex    sync.RWMutex
+
+	// cfg is a copy of the configuration struct that the htlc switch
+	// service was initialized with.
+	cfg *Config
+
+	// pendingPayments is correspondence of user payments and its hashes,
+	// which is used to save the payments which made by user and notify
+	// them about result later.
+	pendingPayments map[lnwallet.PaymentHash][]*pendingPayment
+	pendingMutex    sync.RWMutex
+
+	// circuits is storage for payment circuits which are used to
+	// forward the settle/fail htlc updates back to the add htlc initiator.
+	circuits *circuitMap
+
+	// links is a map of channel id and channel link which manages
+	// this channel.
+	links map[lnwire.ChannelID]*boundedLinkChan
+
+	// forwardCommands is used for propogating the htlc packet forward
+	// requests.
+	forwardCommands chan *forwardPacketCmd
+
+	// chanCloseRequests is used to transfer the channel close request to
+	// the channel close handler.
+	chanCloseRequests chan *ChanClose
+}
+
+// New creates the new instance of htlc switch.
+func New(cfg Config) *Switch {
+	return &Switch{
+		cfg:               &cfg,
+		circuits:          newCircuitMap(),
+		links:             make(map[lnwire.ChannelID]*boundedLinkChan),
+		pendingPayments:   make(map[lnwallet.PaymentHash][]*pendingPayment),
+		forwardCommands:   make(chan *forwardPacketCmd),
+		chanCloseRequests: make(chan *ChanClose),
+		quit:              make(chan struct{}),
+	}
+}
+
+// SendHTLC is used by other subsystems which aren't belong to htlc switch
+// package in order to send the htlc update.
+func (s *Switch) SendHTLC(nextNode []byte, update lnwire.Message) (
+	[sha256.Size]byte, error) {
+
+	htlc := update.(*lnwire.UpdateAddHTLC)
+
+	// Create payment and add to the map of payment in order later to be
+	// able to retrieve it and return response to the user.
+	payment := &pendingPayment{
+		err:         make(chan error, 1),
+		preimage:    make(chan [sha256.Size]byte, 1),
+		paymentHash: htlc.PaymentHash,
+		amount:      htlc.Amount,
+	}
+
+	// Check that we do not have the payment with the same id in order to
+	// prevent map override.
+	s.pendingMutex.Lock()
+	_, ok := s.pendingPayments[htlc.PaymentHash]
```

Comment:
I don't think this is correct. As is now, a repeated payment with the same payment hash (outstanding) will result in an error to the caller. It seems that the reason the map has a slice as it's value is to allow one to iterate over the pending payments, in order to allow asynchronous payments with the same payment hash. 

Alternatively, we could take a stance here at the implementation level, and _disallow_ outstanding payments with repeated payment hashes. This would need to be done in a separate PR though. Possibly in conjunction with implementing sphinx send.